sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1.java

```diff
@@ -189,6 +533,725 @@ private Write() {}
+          JodaClock clock,
+          FirestoreStatefulComponentFactory firestoreStatefulComponentFactory,
+          RpcQosOptions rpcQosOptions) {
+        super(clock, firestoreStatefulComponentFactory, rpcQosOptions);
+      }
+
+      @Override
+      public BatchGetDocuments build() {
+        return genericBuild();
+      }
+
+      @Override
+      BatchGetDocuments buildSafe(
+          JodaClock clock,
+          FirestoreStatefulComponentFactory firestoreStatefulComponentFactory,
+          RpcQosOptions rpcQosOptions) {
+        return new BatchGetDocuments(clock, firestoreStatefulComponentFactory, rpcQosOptions);
+      }
+    }
+  }
+
+  /**
+   * Concrete class representing a {@link PTransform}{@code <}{@link PCollection}{@code <}{@link
+   * PartitionQueryRequest}{@code >, }{@link PTransform}{@code <}{@link RunQueryResponse}{@code >>}
+   * which will read from Firestore.
+   *
+   * <p>This class is part of the Firestore Connector DSL, it has a type safe builder accessible via
+   * {@link FirestoreIO#v1()}{@code .}{@link FirestoreV1#read() read()}{@code .}{@link
+   * FirestoreV1.Read#partitionQuery() partitionQuery()}.
+   *
+   * <p>All request quality-of-service for an instance of this PTransform is scoped to the worker
+   * and configured via {@link PartitionQuery.Builder#withRpcQosOptions(RpcQosOptions)}.
+   *
+   * @see FirestoreIO#v1()
+   * @see FirestoreV1#read()
+   * @see FirestoreV1.Read#partitionQuery()
+   * @see FirestoreV1.PartitionQuery.Builder
+   * @see PartitionQueryRequest
+   * @see RunQueryResponse
+   * @see <a target="_blank" rel="noopener noreferrer"
+   *     href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.Firestore.PartitionQuery">google.firestore.v1.Firestore.PartitionQuery</a>
+   * @see <a target="_blank" rel="noopener noreferrer"
+   *     href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.PartitionQueryRequest">google.firestore.v1.PartitionQueryRequest</a>
+   * @see <a target="_blank" rel="noopener noreferrer"
+   *     href="https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.PartitionQueryResponse">google.firestore.v1.PartitionQueryResponse</a>
+   */
+  public static final class PartitionQuery
+      extends Transform<
+          PCollection<PartitionQueryRequest>,
+          PCollection<RunQueryResponse>,
+          PartitionQuery,
+          PartitionQuery.Builder> {
+
+    private final boolean nameOnlyQuery;
+
+    private PartitionQuery(
+        JodaClock clock,
+        FirestoreStatefulComponentFactory firestoreStatefulComponentFactory,
+        RpcQosOptions rpcQosOptions,
+        boolean nameOnlyQuery) {
+      super(clock, firestoreStatefulComponentFactory, rpcQosOptions);
+      this.nameOnlyQuery = nameOnlyQuery;
+    }
+
+    @Override
+    public PCollection<RunQueryResponse> expand(PCollection<PartitionQueryRequest> input) {
```

Comment:
I'm not fully understanding when this would be used. I would have expected each worker to separately deal with a RunQueryRequest, though maybe there's Beam magic going on here?

Reply:
Agree. I would have expected PartitionQuery to be a utility that help Beam read transform to parallelize better instead being a part of the public API.

Reply:
Given that PartitionQuery is a purely ahead of time operation, it is a transform that happens in the pipeline itself rather than per worker. By doing this, all of the subqueries are resolved and checkpointed before any RunQuery processing is performed thereby allowing the beam runner to absorb the burden of enqueuing all of the individual queries.

We do want PartitionQuery to be part of the public API for composition reasons, and for our power customers who do generate their own PartitionQueryRequests or their processing jobs. This is less error prone than attempting to wrap the PartitionQuery api and offer all of its features rather than simply allowing a pass through of the RPC Proto.

I have pushed a commit where I decoupled PartitionQuery from RunQuery. This change makes the PartitionQuery seem a bit more utility like in that it is now an explicitly added pre-step:

##### Before
```java
   PCollection<PartitionQueryRequest> partitionQueryRequests = ...;
   PCollection<RunQueryResponse> runQueryResponses = partitionQueryRequests
       .apply(FirestoreIO.v1().read().partitionQuery().build());
```

##### After
```java
   PCollection<PartitionQueryRequest> partitionQueryRequests = ...;
   PCollection<RunQueryRequest> runQueryRequests = partitionQueryRequests
       .apply(FirestoreIO.v1().read().partitionQuery().build());
   PCollection<RunQueryResponse> runQueryResponses = runQueryRequests
       .apply(FirestoreIO.v1().read().runQuery().build());
```